util/src/main/java/io/grpc/util/RandomSubsettingLoadBalancer.java

```diff
@@ -0,0 +1,168 @@
+            .setLoadBalancingPolicyConfig(config.childConfig)
+            .build());
+  }
+
+  // implements the subsetting algorithm, as described in A68:
+  // https://github.com/grpc/proposal/pull/423
+  private ResolvedAddresses filterEndpoints(ResolvedAddresses resolvedAddresses, long subsetSize) {
+    // configured subset sizes in the range [Integer.MAX_VALUE, Long.MAX_VALUE] will always fall
+    // into this if statement due to collection indexing limitations in JVM
+    if (subsetSize >= resolvedAddresses.getAddresses().size()) {
+      return resolvedAddresses;
+    }
+
+    ArrayList<EndpointWithHash> endpointWithHashList =
+        new ArrayList<>(resolvedAddresses.getAddresses().size());
+
+    for (EquivalentAddressGroup addressGroup : resolvedAddresses.getAddresses()) {
+      endpointWithHashList.add(
+          new EndpointWithHash(
+              addressGroup,
+              hashFunc.hashString(
+                  addressGroup.getAddresses().get(0).toString(),
+                  StandardCharsets.UTF_8)));
+    }
+
+    Collections.sort(endpointWithHashList, new HashAddressComparator());
+
+    // array is constructed for subset sizes in range [0, Integer.MAX_VALUE), therefore casting
+    // from long to int is not going to overflow here
+    ArrayList<EquivalentAddressGroup> addressGroups = new ArrayList<>((int) subsetSize);
+
+    // for loop is executed for subset sizes in range [0, Integer.MAX_VALUE), therefore indexing
+    // variable is not going to overflow here
+    for (int idx = 0; idx < subsetSize; ++idx) {
+      addressGroups.add(endpointWithHashList.get(idx).addressGroup);
+    }
+
+    return resolvedAddresses.toBuilder().setAddresses(addressGroups).build();
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    switchLb.handleNameResolutionError(error);
+  }
+
+  @Override
+  public void shutdown() {
+    switchLb.shutdown();
+  }
+
+  private static final class EndpointWithHash {
+    public final EquivalentAddressGroup addressGroup;
+    public final HashCode hashCode;
+
+    public EndpointWithHash(EquivalentAddressGroup addressGroup, HashCode hashCode) {
+      this.addressGroup = addressGroup;
+      this.hashCode = hashCode;
+    }
+  }
+
+  private static final class HashAddressComparator implements Comparator<EndpointWithHash> {
+    @Override
+    public int compare(EndpointWithHash lhs, EndpointWithHash rhs) {
+      return BYTE_ARRAY_COMPARATOR.compare(lhs.hashCode.asBytes(), rhs.hashCode.asBytes());
+    }
+  }
+
+  public static final class RandomSubsettingLoadBalancerConfig {
+    public final long subsetSize;
+    public final Object childConfig;
+
+    private RandomSubsettingLoadBalancerConfig(long subsetSize, Object childConfig) {
+      this.subsetSize = subsetSize;
+      this.childConfig = childConfig;
+    }
+
+    public static class Builder {
+      Long subsetSize;
+      Object childConfig;
+
+      public Builder setSubsetSize(Integer subsetSize) {
```

Comment:
Use a primitive `int` since it can't be null? Seems that would be more straight-forward. I see that the primary caller has an Integer and it will be stored as a Long. Really, we don't need to store it as `Long` as the value isn't allowed to be zero or negative, so we could detect unset without boxing.